File: StaticSharp/CrossplatformLibrary/Storage/Storage.cs
using Javascriptifier;
using StaticSharp.Gears;
using System;
using System.Linq.Expressions;

namespace StaticSharp {
    namespace Js {

        public static class Storage {

            [JavascriptOnlyMember]
            public static T Restore<T>(string name, Expression<Func<T>> getter) where T: struct => throw new Javascriptifier.JavascriptOnlyException();
        }

    }


}

File: StaticSharp/CrossplatformLibrary/Storage/Storage.js
var Storage = {}

Storage.Elements = {}

Storage.Restore = function (name, getter) {
    let storage = sessionStorage

    let valueString = storage.getItem(name)

    if (Storage.Elements[name] == undefined) {
        Storage.Elements[name] = new Reaction(() => {
            let newValue = getter()
            storage.setItem(name, typeof (newValue) + " " + newValue)
        })
    }

    if (valueString != null) {
        let seperator = valueString.indexOf(' ')
        let type = valueString.substring(0, seperator)
        let value = valueString.substring(seperator + 1)

        if (type == "undefined") {
            return undefined
        }
        if (type == "number") {
            return Number(value)
        }
        if (type == "boolean") {
            return (value === 'true')
        }
        if (type == "string") {
            return value
        }

    } else {
        return undefined
    }

    

}